presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java

```diff
@@ -0,0 +1,517 @@
+    }
+
+    @Override
+    public long getRegionLogicalSizeInBytes(int position, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        int entriesStart = getOffsets()[getOffsetBase() + position];
+        int entriesEnd = getOffsets()[getOffsetBase() + position + length];
+        int entryCount = entriesEnd - entriesStart;
+
+        return getRawKeyBlock().getRegionLogicalSizeInBytes(entriesStart, entryCount) +
+                getRawValueBlock().getRegionLogicalSizeInBytes(entriesStart, entryCount) +
+                (Integer.BYTES + Byte.BYTES) * (long) length +
+                Integer.BYTES * HASH_MULTIPLIER * (long) entryCount;
+    }
+
+    @Override
+    public long getApproximateRegionLogicalSizeInBytes(int position, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        int entriesStart = getOffset(position);
+        int entriesEnd = getOffset(position + length);
+        int entryCount = entriesEnd - entriesStart;
+
+        return getRawKeyBlock().getApproximateRegionLogicalSizeInBytes(entriesStart, entryCount) +
+                getRawValueBlock().getApproximateRegionLogicalSizeInBytes(entriesStart, entryCount) +
+                (Integer.BYTES + Byte.BYTES) * (long) length +         // offsets and mapIsNull
+                Integer.BYTES * HASH_MULTIPLIER * (long) entryCount;   // hash tables
+    }
+
+    @Override
+    public long getPositionsSizeInBytes(boolean[] positions)
+    {
+        // We can use either the getRegionSizeInBytes or getPositionsSizeInBytes
+        // from the underlying raw blocks to implement this function. We chose
+        // getPositionsSizeInBytes with the assumption that constructing a
+        // positions array is cheaper than calling getRegionSizeInBytes for each
+        // used position.
+        int positionCount = getPositionCount();
+        checkValidPositions(positions, positionCount);
+        boolean[] entryPositions = new boolean[getRawKeyBlock().getPositionCount()];
+        int usedEntryCount = 0;
+        int usedPositionCount = 0;
+        for (int i = 0; i < positions.length; ++i) {
+            if (positions[i]) {
+                usedPositionCount++;
+                int entriesStart = getOffsets()[getOffsetBase() + i];
+                int entriesEnd = getOffsets()[getOffsetBase() + i + 1];
+                for (int j = entriesStart; j < entriesEnd; j++) {
+                    entryPositions[j] = true;
+                }
+                usedEntryCount += (entriesEnd - entriesStart);
+            }
+        }
+        return getRawKeyBlock().getPositionsSizeInBytes(entryPositions) +
+                getRawValueBlock().getPositionsSizeInBytes(entryPositions) +
+                (Integer.BYTES + Byte.BYTES) * (long) usedPositionCount +
+                Integer.BYTES * HASH_MULTIPLIER * (long) usedEntryCount;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).